metagraph/src/common/vectors/vector_algorithm.hpp

```diff
@@ -109,6 +110,184 @@ inline void unset_bit(uint64_t *v,
+                             int memorder = __ATOMIC_SEQ_CST) {
+    size_t width = vector.width();
+    size_t bit_pos = i * width;
+    uint64_t mask = (1llu << width) - 1;
+    if (width + 7 > 64) {
+        // there is no way to reliably modify without a mutex
+        std::lock_guard<std::mutex> lock(backup_mutex);
+        return vector[i];
+    } else if ((bit_pos & 0x3F) + width <= 64) {
+        // element fits in an aligned word
+        const uint64_t *word = &vector.data()[bit_pos >> 6];
+        return (__atomic_load_n(word, memorder) >> (bit_pos & 0x3F)) & mask;
+    } else if ((bit_pos & 0x7F) + width <= 128) {
+        // element fits in an aligned double word
+        const __uint128_t *word = &reinterpret_cast<const __uint128_t*>(vector.data())[bit_pos >> 7];
+        return (__atomic_load_n(word, memorder) >> (bit_pos & 0x7F)) & mask;
+    } else {
+        uint8_t shift = bit_pos & 0x7;
+        const uint8_t *word = &reinterpret_cast<const uint8_t*>(vector.data())[bit_pos >> 3];
+        if (shift + width <= 8) {
+            // read from a byte
+            return (__atomic_load_n(word, memorder) >> shift) & mask;
+        } else if (shift + width <= 16) {
+            // unaligned read from two bytes
+            return (__atomic_load_n((const uint16_t*)word, memorder) >> shift) & mask;
+        } else if (shift + width <= 32) {
+            // unaligned read from four bytes
+            return (__atomic_load_n((const uint32_t*)word, memorder) >> shift) & mask;
+        } else if (shift + width <= 64) {
+            // unaligned read from eight bytes
+            return (__atomic_load_n((const uint64_t*)word, memorder) >> shift) & mask;
+        } else {
+            assert(false);
+        }
+    }
+
+    return 0;
+}
+
+inline uint64_t atomic_fetch_and_add(sdsl::int_vector<> &vector, uint64_t i,
+                                     uint64_t count,
+                                     std::mutex &backup_mutex,
+                                     int memorder = __ATOMIC_SEQ_CST) {
+    size_t width = vector.width();
+    size_t bit_pos = i * width;
+    uint64_t mask = (1llu << width) - 1;
+    if (width + 7 > 64) {
+        // there is no way to reliably modify without a mutex
+        std::lock_guard<std::mutex> lock(backup_mutex);
+        uint64_t old_val = vector[i];
+        vector[i] += count;
+        return old_val;
+    } else if ((bit_pos & 0x3F) + width <= 64) {
+        // element fits in an aligned word
+        uint64_t *word = &vector.data()[bit_pos >> 6];
+        uint8_t shift = bit_pos & 0x3F;
+        return (__atomic_fetch_add(word, count << shift, memorder) >> shift) & mask;
+    } else if ((bit_pos & 0x7F) + width <= 128) {
+        // element fits in an aligned double word
+        __uint128_t *word = &reinterpret_cast<__uint128_t*>(vector.data())[bit_pos >> 7];
+        uint8_t shift = bit_pos & 0x7F;
+        return (__sync_fetch_and_add(word, __uint128_t(count) << shift) >> shift) & mask;
```

Comment:
Add comment here that gcc only generates a `cmpxchg16b` instruction when older `__sync` instructions are used...